src/main.rs

```diff
@@ -93,46 +102,299 @@ macro_rules! create_socket {
+
+        cp.SCB.enable_icache();
 
         init_log();
         // info!("Version {} {}", build_info::PKG_VERSION, build_info::GIT_VERSION.unwrap());
         // info!("Built on {}", build_info::BUILT_TIME_UTC);
         // info!("{} {}", build_info::RUSTC_VERSION, build_info::TARGET);
 
-        // c.schedule.tick(Instant::now()).unwrap();
+        let mut debug_pin = gpioc.pc6.into_push_pull_output();
+        debug_pin.set_low().unwrap();
+
+        let mut dac_pin = gpiob.pb15.into_push_pull_output();
+        dac_pin.set_low().unwrap();
+
+        // Configure timer 2 to trigger conversions for the ADC
+        let mut timer2 = dp.TIM2.timer(500.khz(), &mut clocks);
+        timer2.listen(hal::timer::Event::TimeOut);
 
-        let dp = c.device;
         init::LateResources {
-            spi: (dp.SPI1, dp.SPI2, dp.SPI4, dp.SPI5),
-            i2c: dp.I2C2,
-            ethernet_periph: (
-                dp.ETHERNET_MAC,
-                dp.ETHERNET_DMA,
-                dp.ETHERNET_MTL,
-            ),
+            adc1: adc1_spi,
+            dac1: dac1_spi,
+            adc2: adc2_spi,
+            dac2: dac2_spi,
+
+            dbg_pin: debug_pin,
+            dac_pin: dac_pin,
+            timer: timer2,
+
+            _eeprom_i2c: i2c2,
+//            ethernet_periph: (
+//                dp.ETHERNET_MAC,
+//                dp.ETHERNET_DMA,
+//                dp.ETHERNET_MTL,
+//            ),
+        }
+    }
+
+    #[task(binds = TIM2, resources = [dbg_pin, timer, adc1, adc2])]
+    fn tim2(mut c: tim2::Context) {
+        c.resources.timer.clear_uif_bit();
+        c.resources.dbg_pin.set_high().unwrap();
+
+        // Start a SPI transaction on ADC0 and ADC1
+        c.resources.adc1.lock(|adc| adc.spi.cr1.modify(|_, w| w.cstart().set_bit()));
+        c.resources.adc2.lock(|adc| adc.spi.cr1.modify(|_, w| w.cstart().set_bit()));
+
+        c.resources.dbg_pin.set_low().unwrap();
+    }
+
+    #[task(binds = SPI2, resources = [adc1, dac1, adc2, dac2, iir_state, iir_ch, dac_pin], priority = 2)]
+    fn adc_spi(c: adc_spi::Context) {
+        #[cfg(feature = "bkpt")]
+        cortex_m::asm::bkpt();
+
+        c.resources.dac_pin.set_high().unwrap();
+
+        let output_ch1 = {
+            let a: u16 = c.resources.adc1.read().unwrap();
+            let x0 = f32::from(a as i16);
+            let y0 = c.resources.iir_ch[0].update(&mut c.resources.iir_state[0], x0);
+            y0 as i16 as u16 ^ 0x8000
+        };
+        c.resources.adc1.spi.ifcr.write(|w| w.eotc().set_bit());
+
+        let output_ch2 = {
+            let a: u16 = nb::block!(c.resources.adc2.read()).unwrap();
+            let x0 = f32::from(a as i16);
+            let y0 = c.resources.iir_ch[1].update(&mut c.resources.iir_state[1], x0);
+            y0 as i16 as u16 ^ 0x8000
+        };
+        c.resources.adc2.spi.ifcr.write(|w| w.eotc().set_bit());
+
+        c.resources.dac1.send(output_ch1).unwrap();
```

Comment:
This now has different latency adc-dac for both channels. Important to write to the DACs as early as possible.

Reply:
This actually is done to make an effort to ensure the channel latency is nearly identical for each output. Because the sampling is triggered near instantaneously for each channel, the theory here is to generate the outputs simultaneously as well, which will result in a matched latency. If you update CH1 first, CH2 will have a latency of 400uS longer while it performs the IIR filter.

Reply:
Sure.
Lower latency (even on only one channel) has priority over matched latency.
The better solution for the future will be to space the two ADC xfers by about one IIR processing time so that they both have matched and low latency. (Would need two channels on that timer).
I think that "uS" should be "ns". And is it really that much? I seem to remember less.

Reply:
You are correct in the uS being nS. I'll delve more into the IIR filter time using the cycle counter later to verify

Reply:
This has been corrected by updating the two channels to occur at a 50% duty cycle offset.